### ISSUE TITLE
Online while  not

### ISSUE THREAD

**Issue by Ana-Green** [HalcyonGrid/halcyon#100]
[21:21] ZHAO Bento AO Victor: 709% memory free
[21:21] Ana Green is online.
21:22] Grid Owner: Hi Ana 
[21:22] Ana Green: User is not logged in. Message saved.
![online but not](https://user-images.githubusercontent.com/36156774/77840326-7fafb780-7186-11ea-84dd-c39645a03161.png)

**Comment by emperorstarfinder** [comment]
Yes, this is a bug.  I just haven't had a chance to look into what may be missing or incorrect in the code at yet.  But it has been confirmed on my team's internal testbed grid.

**Comment by Ana-Green** [comment]
ok  thank  you emperorstarfinder 

Regards,
Ana Green

**Comment by appurist** [comment]
This happens if a user does not exit cleanly, such as when a viewer crashes. The code that clears the record in the `agents` table runs as part of logoff, rather than as part of the disconnect code, so it does not run. If I recall correctly, I think some attempts were made to clean this up correctly from disconnects, but that there was still a case where the `agents` record remained, so the grid thinks the user is still logged in.

This problem also has a nasty ripple effect because some code, such as IM delivery, checks first to see if the user is online. This means that sometimes you won't get offline IMs delivered in a viewer until the _second_ login after it is sent.  And IM delivery includes inventory deliveries, friend requests, etc.

A workaround for the user in question is to always log back in and out again after a crash, but that's easy to forget or not bother with and doesn't change the fact that there is still a cleanup bug in the Halcyon code. I'm just mentioning this because sometimes it helps.
